src/deferrable.c

```diff
@@ -0,0 +1,210 @@
+#include <deferrable.h>
+#include <string.h>
+#include <stdio.h>
+
+static rt_uint8_t active;
+
+/**
+ * This function will start a new period of the deferrable server
+ *
+ * @param NULL parameter passed to the timer function
+ *
+ * @return void
+ */
+void rt_timer_deferrable_period(void *param)
+{
+    deferrable_server.capacity_left = deferrable_server.capacity;
+
+    if(rt_list_isempty(deferrable_server.asyncTask.next))
+        rt_thread_resume(&deferrable_server.deferrable);
+
+    rt_timer_control(&deferrable_server.tim_capacity, RT_TIMER_CTRL_SET_TIME, &deferrable_server.capacity_left);
+}
+
+/**
+ * This function will stop the deferrable server when its capacity is finished
+ *
+ * @param NULL parameter passed to the timer function
+ *
+ * @return void
+ */
+void rt_timer_deferrable_capacity(void *param)
+{
+    deferrable_server.capacity_left = 0;
+    rt_thread_suspend(&deferrable_server.deferrable);
```

Comment:
Do you want suspend other thread or suspend the thread itself?
rt-thread's suspend API only can be invoked in a thread and suspend this thread itself. You cannot invoke rt_thread_suspend API

Reply:
This is invoked only while the server is the current running thread, so it suspends itself

Reply:
Nop, it cannot be invoked by current thread. Timers will run in another thread or directly run in interruptes.

Reply:
Okay, I understand, how can this be done instead? Is there a way to do something similar from another thread